test/LocalVocabTest.cpp

```diff
@@ -20,23 +34,146 @@ TEST(LocalVocab, construction) {
+  // Query execution context (with small test index) and allocator for testing,
+  // see `IndexTestHelpers.h`.
+  using ad_utility::AllocatorWithLimit;
+  QueryExecutionContext* testQec = ad_utility::testing::getQec();
+  AllocatorWithLimit<Id> testAllocator = ad_utility::testing::makeAllocator();
+
+  // Lambda that checks the contents of the local vocabulary after the specified
+  // operation.
+  // NOTE: Clear the cache before each check
+  // TODO: Does this solve the problem with the local vocab when doing several
+  // operations on `values1`?
+  auto checkLocalVocab = [&](auto* operation,
+                             std::vector<std::string> expectedWords) -> void {
+    operation->getExecutionContext()->getQueryTreeCache().clearAll();
+    ResultTable resultTable(testAllocator);
+    operation->computeResult(&resultTable);
+    std::vector<std::string> localVocabWords;
+    for (size_t i = 0; i < resultTable._localVocab->size(); ++i) {
+      localVocabWords.emplace_back(
+          resultTable._localVocab->getWord(LocalVocabIndex::make(i)));
+    }
+    ASSERT_EQ(localVocabWords, expectedWords)
+        << "Operation: " << operation->getDescriptor() << std::endl;
+  };
+
+  // Lambda that checks that `computeResult` throws an exception.
+  auto checkThrow = [&](auto* operation) -> void {
+    ResultTable resultTable(testAllocator);
+    ASSERT_THROW(operation->computeResult(&resultTable), std::runtime_error)
+        << "Operation: " << operation->getDescriptor() << std::endl;
+  };
+
+  // Lambda that returns a `std::shared_ptr` to a `QueryExecutionTree` with only
+  // the given operation (and using `testQec`).
+  auto qet = [&](const auto& operation) -> std::shared_ptr<QueryExecutionTree> {
+    return std::make_shared<QueryExecutionTree>(
+        testQec, std::make_shared<Values>(operation));
+  };
+
+  // VALUES operation with two variables and two rows. Adds four new literals.
+  Values values1(testQec, {{Variable{"?x"}, Variable{"?y"}},
+                           {{TripleComponent{"x"}, TripleComponent{"y1"}},
+                            {TripleComponent{"x"}, TripleComponent{"y2"}}}});
+  checkLocalVocab(&values1, std::vector<std::string>{"x", "y1", "y2"});
+
+  // VALUES operation that uses an existing literal (from the test index). Note
+  // that the quotes are part of the name of the literals.
+  Values values2(testQec, {{Variable{"?x"}}, {{TripleComponent{"\"alpha\""}}}});
+  checkLocalVocab(&values2, std::vector<std::string>{});
+
+  // JOIN operation with exactly one non-empty local vocab.
+  Join join1(testQec, qet(values1), qet(values2), 1, 1);
+  checkLocalVocab(&join1, std::vector<std::string>{"x", "y1", "y2"});
+
+  // JOIN operation with two non-empty local vocab.
+  Join join2(testQec, qet(values1), qet(values1), 1, 1);
+  checkThrow(&join2);
+
+  // ORDER BY operation.
+  // Note: the third argument ar the indices of the columns to be sorted, and
+  // the sort order. It's not important for this test.
```

Comment:
It's important, that all the indices are smaller than the number of variables in the subtree, 
but `0` and `1` are ok, since you have two variables in `values1`.